File: matlab/utils/lib/DMP_lib/DMP_orient.m
%% DMP_orient class
%  For encoding Cartesian Orientation.
%


classdef DMP_orient < handle
    
    properties (Constant)
        
          %% enum ARGUMENT
          
          %% enum TRAIN_METHOD
          LWR = 201;
          LS = 203;

    end
       
    properties
        
        Qgd % Trained target orientation as unit quaternion.
        Q0d % Trained initial orientation as unit quaternion.
        log_Qgd_invQ0d % log(Qgd * inv(Q0d)), precalculated
        tau_d % Trained time-scaling.
        
        eq_f % 3x1 vector of WSoG producing the desired orientation.
        vRot_f % 3x1 vector of WSoG producing the desired rotational velocity.
        dvRot_f % 3x1 vector of WSoG producing the desired rotational acceleration.
        
        a_z % parameter 'a_z' relating to the spring-damper system
        b_z % parameter 'b_z' relating to the spring-damper system

        can_clock_ptr % handle (pointer) to the canonical clock
        shape_attr_gating_ptr % pointer to gating function for the shape attractor

        zero_tol % tolerance value used to avoid divisions with very small numbers

    end

    methods
        %% DMP_orient constructor.
        %  @param[in] N_kernels: 3x3 matrix where each row contains the kernels for position,
        %                       velocity and acceleration and each row for x, y and z coordinate.
        %                       If 'N_kernels' is a column vector then every coordinate will have the
        %                       same number of kernels.
        %  @param[in] a_z: Parameter 'a_z' relating to the spring-damper system.
        %  @param[in] b_z: Parameter 'b_z' relating to the spring-damper system.
        %  @param[in] can_clock_ptr: Pointer to a DMP canonical system object.
        %  @param[in] shape_attr_gating_ptr: Pointer to a DMP gating function object.
        %
        function this = DMP_orient(N_kernels, a_z, b_z, can_clock_ptr, shape_attr_gating_ptr)

            this.zero_tol = 1e-30; %realmin;
            this.a_z = a_z;
            this.b_z = b_z;
            this.can_clock_ptr = can_clock_ptr;
            this.shape_attr_gating_ptr = shape_attr_gating_ptr;

            if (isscalar(N_kernels)), N_kernels = ones(3,1)*N_kernels; end
            if (size(N_kernels,2) == 1), N_kernels = repmat(N_kernels,1,3); end

            shapeAttrGatFun_ptr = @(x)getOutput(this.shape_attr_gating_ptr,x);
            for i=1:3
                this.eq_f{i} = WSoG(N_kernels(1,i), shapeAttrGatFun_ptr);
                this.vRot_f{i} = WSoG(N_kernels(2,i), shapeAttrGatFun_ptr);
                this.dvRot_f{i} = WSoG(N_kernels(3,i), shapeAttrGatFun_ptr);
            end

        end
        
        
        %% Trains the DMP_orient.
        %  @param[in] train_method: The training method (see dmp_::TRAIN_METHOD enum).
        %  @param[in] Time: 1xN row vector with the timestamps of the training data points.
        %  @param[in] Qd_data: 4xN matrix with desired orientation as unit quaternion at each column for each timestep.
        %  @param[in] vRotd_data: 3xN matrix with desired angular velocity at each column for each timestep.
        %  @param[in] dvRotd_data: 3xN matrix with desired angular acceleration at each column for each timestep.
        %  @param[in] train_error: Optinal pointer to return the training error as norm(F-Fd)/n_data.
        %
        function train_err = train(this, train_method, Time, Qd_data, vRotd_data, dvRotd_data)

            n_data = length(Time);
            
            this.Q0d = Qd_data(:,1);
            this.Qgd = Qd_data(:,end);
            this.log_Qgd_invQ0d = this.quatLog( this.quatProd(this.Qgd, this.quatInv(this.Q0d) ) );
            
            eqd = zeros(3, n_data);           
            for j=1:n_data, eqd(:,j) = this.quatLog(this.quatProd(Qd_data(:,j),this.quatInv(this.Qgd))); end

            tau = Time(end);
            this.tau_d = tau;
            this.setTau(tau);
    
            x = this.phase(Time);
            
            if (train_method == DMP_orient.LWR), wsog_train_method = WSoG.LWR;
            elseif (train_method == DMP_orient.LS), wsog_train_method = WSoG.LS;  
            else, error('[DMP_orient::train]: Unsopported training method...');
            end
                
            if (nargout > 0)
                train_err = zeros(3,3);
                for i=1:3
                    train_err(1,i) = this.eq_f{i}.train(wsog_train_method, x, eqd(i,:));
                    train_err(2,i) = this.vRot_f{i}.train(wsog_train_method, x, vRotd_data(i,:));
                    train_err(3,i) = this.dvRot_f{i}.train(wsog_train_method, x, dvRotd_data(i,:));
                end
            else
                for i=1:3
                    this.eq_f{i}.train(wsog_train_method, x, eqd(i,:));
                    this.vRot_f{i}.train(wsog_train_method, x, vRotd_data(i,:));
                    this.dvRot_f{i}.train(wsog_train_method, x, dvRotd_data(i,:));
                end
            end

        end
        
        
        %% Returns the angular acceleration for the given input state defined by the timestamp,
        %  the orientation, the angular velocity and acceleration, the initial and target orientation
        %  and an optinal coupling term.
        %  @param[in] x: phase variable.
        %  @param[in] Q: Current orientation as unit quaternion.
        %  @param[in] vRot: Current rotational velocity.
        %  @param[in] Q0: initial orientation as unit quaternion.
        %  @param[in] Qg: Goal orientation as unit quaternion.
        %  @param[in] y_c: Coupling term. (optional, default=arma::vec().zeros(3))
        %  @return dvRot: Rotational acceleration.
        %
        function dvRot = getRotAccel(this, x, Q, vRot, Q0, Qg, y_c)

            if (nargin < 7), y_c=zeros(3,1); end

            eqd = zeros(3,1);
            vRotd = zeros(3,1);
            dvRotd = zeros(3,1);
            
            for i=1:3
                eqd(i) = this.eq_f{i}.output(x);
                vRotd(i) = this.vRot_f{i}.output(x);
                dvRotd(i) = this.dvRot_f{i}.output(x);
            end 

            Qd = this.quatProd( this.quatExp(eqd), this.Qgd );
            
            tau = this.getTau();
            kt = this.tau_d / tau;
            ks = this.quatLog( this.quatProd(Qg, this.quatInv(Q0) ) ) ./ this.log_Qgd_invQ0d;
            
            QQg = this.quatProd(Q,this.quatInv(Qg));
            inv_exp_QdQgd = this.quatInv( this.quatExp( ks.* this.quatLog( this.quatProd(Qd,this.quatInv(this.Qgd)) ) ) );
            
            eqd = eqd'
            vRotd = vRotd'
            dvRotd = dvRotd'
            Qd = Qd'
            tau
            kt
            ks = ks'
            QQg = QQg'
            inv_exp_QdQgd = inv_exp_QdQgd'
            error('Stop');
            
            dvRot = kt^2*ks.*dvRotd - (this.a_z/tau)*(vRot-kt*ks.*vRotd) ...
                    -(this.a_z*this.b_z/tau^2) * this.quatLog ( this.quatProd(QQg, inv_exp_QdQgd)) + y_c;
            
        end
        

        %% Returns the time scaling factor.
        %  @return: The time scaling factor.
        %
        function tau = getTau(this)

            tau = this.can_clock_ptr.getTau();

        end
        
        
        %% Sets the time scaling factor.
        %  @param[in] tau: The time scaling factor.
        %
        function setTau(this, tau)

            this.can_clock_ptr.setTau(tau);

        end
        
        
        %% Returns the phase variable corresponding to the given time instant.
        %  @param[in] t: The time instant.
        %  @return: The phase variable for time 't'.
        %
        function x = phase(this, t)
            
            x = this.can_clock_ptr.getPhase(t);

        end
        
        
        %% Returns the derivative of the phase variable.
        %  @param[in] x: The phase variable.
        %  @return: The derivative of the phase variable.
        %
        function dx = phaseDot(this, x)
            
            dx = this.can_clock_ptr.getPhaseDot(x);

        end

    end
    
    methods (Static)
        
        function v_rot = quatLog(Q)

            n = Q(1);
            e = Q(2:4);
            norm_e = norm(e);

            if (norm_e > 1e-16)
                theta = 2*real(atan2(norm_e,n));
                v_rot = theta*e/norm_e;
            else
                v_rot = zeros(size(e));
            end

        end
        
        function Q = quatExp(v_rot)

            norm_v_rot = norm(v_rot);
            theta = norm_v_rot;

            if (norm_v_rot > 1e-16)
                Q(1) = cos(theta/2);
                Q(2:4) = sin(theta/2)*v_rot/norm_v_rot;
            else
                Q = [1 0 0 0]';
            end

        end
        
        function Q12 = quatProd(Q1, Q2)

            Q1 = Q1(:);
            Q2 = Q2(:);

            n1 = Q1(1);
            e1 = Q1(2:4);

            n2 = Q2(1);
            e2 = Q2(2:4);

            Q12 = zeros(4,1);
            Q12(1) = n1*n2 - e1'*e2;
            Q12(2:4) = n1*e2 + n2*e1 + cross(e1,e2);

        end
        
        function invQ = quatInv(Q)

            invQ = zeros(size(Q));

            invQ(1) = Q(1);
            invQ(2:4) = -Q(2:4);

        end

    end
end

File: c++/src/lib/dmp_lib/src/DMP/DMP_orient.cpp
#include <dmp_lib/DMP/DMP_orient.h>

namespace as64_
{

namespace dmp_
{

  arma::vec quatLog(const arma::vec &quat, double zero_tol=1e-16)
  {
    arma::vec e = quat.subvec(1,3);
    double n = quat(0);

    if (n > 1) n = 1;
    if (n < -1) n = -1;

    arma::vec omega(3);
    double e_norm = arma::norm(e);

    if (e_norm > zero_tol) omega = 2*std::atan2(e_norm,n)*e/e_norm;
    else omega = arma::vec().zeros(3);

    return omega;
  }

  arma::vec quatExp(const arma::vec &v_rot, double zero_tol=1e-16)
  {
    arma::vec quat(4);
    double norm_v_rot = arma::norm(v_rot);
    double theta = norm_v_rot;

   if (norm_v_rot > zero_tol)
   {
      quat(0) = std::cos(theta/2);
      quat.subvec(1,3) = std::sin(theta/2)*v_rot/norm_v_rot;
    }
    else{
      quat << 1 << 0 << 0 << 0;
    }

    return quat;
  }

  arma::vec quatProd(const arma::vec &quat1, const arma::vec &quat2)
  {
    arma::vec quat12(4);

    double n1 = quat1(0);
    arma::vec e1 = quat1.subvec(1,3);

    double n2 = quat2(0);
    arma::vec e2 = quat2.subvec(1,3);

    quat12(0) = n1*n2 - arma::dot(e1,e2);
    quat12.subvec(1,3) = n1*e2 + n2*e1 + arma::cross(e1,e2);

    return quat12;
  }

  arma::vec quatInv(const arma::vec &quat)
  {
    arma::vec quatI(4);

    quatI(0) = quat(0);
    quatI.subvec(1,3) = - quat.subvec(1,3);

    return quatI;
  }

} // namespace dmp_


DMP_orient::DMP_orient(const arma::Mat<int> &N_kernels, double a_z, double b_z, std::shared_ptr<CanonicalClock> &can_clock_ptr,
  std::shared_ptr<GatingFunction> &shape_attr_gating_ptr)
{
  this->zero_tol = 1e-30; // realmin;
  this->a_z = a_z;
  this->b_z = b_z;
  this->can_clock_ptr = can_clock_ptr;
  this->shape_attr_gating_ptr = shape_attr_gating_ptr;

  std::function<double(double)> shapeAttrGatFun_ptr =
  std::bind(static_cast<double(GatingFunction::*)(double)const>(&GatingFunction::getOutput),
            shape_attr_gating_ptr.get(), std::placeholders::_1);

  arma::Mat<int> n_ker(3,3);
  if (N_kernels.n_cols==1)
  {
    for (int i=0; i<3; i++) n_ker(i,0)=n_ker(i,1)=n_ker(i,2)=N_kernels(i);
  }

  this->eq_f.resize(3);
  this->vRot_f.resize(3);
  this->dvRot_f.resize(3);
  for (int i=0; i<3; i++)
  {
      this->eq_f[i].reset(new dmp_::WSoG(n_ker(0,i), shapeAttrGatFun_ptr));
      this->vRot_f[i].reset(new dmp_::WSoG(n_ker(1,i), shapeAttrGatFun_ptr));
      this->dvRot_f[i].reset(new dmp_::WSoG(n_ker(2,i), shapeAttrGatFun_ptr));
  }
}


void DMP_orient::train(dmp_::TRAIN_METHOD train_method, const arma::rowvec &Time, const arma::mat &Qd_data,
                      const arma::mat &vRotd_data, const arma::mat &dvRotd_data, arma::mat *train_err)
{
    int n_data = Time.size();
    int i_end = n_data-1;

    this->Q0d = Qd_data.col(0);
    this->Qgd = Qd_data.col(i_end);
    this->log_Qgd_invQ0d = dmp_::quatLog( dmp_::quatProd(this->Qgd, dmp_::quatInv(this->Q0d) ) );

    arma::mat eqd(3, n_data);
    for (int j=0; j<n_data; j++) eqd.col(j) = dmp_::quatLog(dmp_::quatProd(Qd_data.col(j),dmp_::quatInv(this->Qgd)));

    double tau = Time(i_end);
    this->tau_d = tau;
    this->setTau(tau);

    arma::rowvec x(n_data);
    for (int j=0; j<n_data; j++) x(j) = this->phase(Time(j));

    if (train_err)
    {
      train_err->resize(3,3);
      for (int i=0; i<3; i++)
      {
          this->eq_f[i]->train(train_method, x, eqd.row(i), &(train_err->at(0,i)));
          this->vRot_f[i]->train(train_method, x, vRotd_data.row(i), &(train_err->at(1,i)));
          this->dvRot_f[i]->train(train_method, x, dvRotd_data.row(i), &(train_err->at(2,i)));
      }
    }
    else
    {
      for (int i=0; i<3; i++)
      {
          this->eq_f[i]->train(train_method, x, eqd.row(i));
          this->vRot_f[i]->train(train_method, x, vRotd_data.row(i));
          this->dvRot_f[i]->train(train_method, x, dvRotd_data.row(i));
      }
    }
}


arma::vec DMP_orient::getRotAccel(double x, const arma::vec &Q, const arma::vec &vRot,
            const arma::vec &Q0, const arma::vec &Qg, const arma::vec &Z_c)
{

    arma::vec eqd(3);
    arma::vec vRotd(3);
    arma::vec dvRotd(3);
    for (int i=0; i<3; i++)
    {
        eqd(i) = this->eq_f[i]->output(x);
        vRotd(i) = this->vRot_f[i]->output(x);
        dvRotd(i) = this->dvRot_f[i]->output(x);
    }

    arma::vec Qd = dmp_::quatProd( dmp_::quatExp(eqd), this->Qgd );

    double tau = this->getTau();
    double kt = this->tau_d / tau;
    arma::vec ks = dmp_::quatLog( dmp_::quatProd(Qg, dmp_::quatInv(Q0) ) ) / this->log_Qgd_invQ0d;

    arma::vec QQg = dmp_::quatProd(Q,dmp_::quatInv(Qg));
    arma::vec inv_exp_QdQgd = dmp_::quatInv( dmp_::quatExp( ks % dmp_::quatLog( dmp_::quatProd(Qd,dmp_::quatInv(this->Qgd)) ) ) );

    arma::vec dvRot = std::pow(kt,2)*ks%dvRotd - (this->a_z/tau)*(vRot-kt*ks%vRotd)
            -(this->a_z*this->b_z/std::pow(tau,2)) * dmp_::quatLog ( dmp_::quatProd(QQg, inv_exp_QdQgd)) + Z_c;

    return dvRot;
}


double DMP_orient::getTau() const
{
    return this->can_clock_ptr->getTau();
}


void DMP_orient::setTau(double tau)
{
    this->can_clock_ptr->setTau(tau);
}


double DMP_orient::phase(double t) const
{
    return this->can_clock_ptr->getPhase(t);
}


double DMP_orient::phaseDot(double x) const
{
    return this->can_clock_ptr->getPhaseDot(x);
}

} // namespace as64_
